Add Bitmap tests

hughbe opened the pull request:
50-50 cleanup from Mono and handwritten tests

hughbe commented:
This will conflict with all the other PRs due to changes in the csproj file :(
But, this PR relies on #20811

mellinoe commented:
Hey @hughbe, been on vacation for the past few days. Sorry for the delay. I will look at this in the next couple days along with other Drawing stuff.

mellinoe commented on src/System.Drawing.Common/tests/BitmapTests.cs:
```diff
@@ -0,0 +1,1531 @@
+
+                Bitmap deserializedBitmap = (Bitmap)formatter.Deserialize(memoryStream);
+                Assert.Equal(bitmap.Size, deserializedBitmap.Size);
+                Assert.Equal(
```
I would rather just see `data[0]`, `data[1]`, etc. rather than have this index variable being incremented inline.

mellinoe replied:
Why construct `red` and `blue` like this?

hughbe replied:
I replaced this with
```
Color red = Color.FromArgb(Color.Red.ToArgb());
Color blue = Color.FromArgb(Color.Blue.ToArgb());
```

This is because `Color.Red` is actually not equal to `Color.FromArgb(Color.Red.ToArgb());`

Color.Red is a "known color", but known color metadata is stripped in calls to `SetPixel` or `GetPixel`

hughbe replied:
I've updated it in some places. In others, it can't go, such as in the following code, as the index depends on the size
```cs
int n = 0;
Assert.Equal(c.B, data[n++]);
Assert.Equal(c.G, data[n++]);
Assert.Equal(c.R, data[n++]);
if (size % 4 == 0)
{
    if (format == PixelFormat.Format32bppRgb)
    {
        Assert.Equal(128, data[n++]);
    }
    else
    {
        Assert.Equal(c.A, data[n++]);
    }
}
Assert.Equal(d.B, data[n++]);
Assert.Equal(d.G, data[n++]);
Assert.Equal(d.R, data[n++]);
if (size % 4 == 0)
{
    if (format == PixelFormat.Format32bppRgb)
    {
        Assert.Equal(192, data[n++]);
    }
    else
    {
        Assert.Equal(d.A, data[n++]);
    }
}
```

hughbe commented on src/System.Drawing.Common/tests/BitmapTests.cs:
```diff
@@ -0,0 +1,1531 @@
+            Assert.Equal(expectedReserved, data.Reserved);
+
+            // Unlocking 16bppGrayscale fails.
+            try
```
Thanks, fixed

hughbe commented on src/System.Drawing.Common/tests/BitmapTests.cs:
```diff
@@ -0,0 +1,1531 @@
+        [MemberData(nameof(Image_TestData))]
+        public void Ctor_Width_Height_Graphics(Bitmap image, int width, int height)
+        {
+            Graphics graphics = graphics = Graphics.FromImage(image);
```
Thanks, fixed

hughbe commented on src/System.Drawing.Common/tests/BitmapTests.cs:
```diff
@@ -0,0 +1,1531 @@
+        [MemberData(nameof(Ctor_FilePath_TestData))]
+        public void Ctor_FilePath(string filename, int width, int height, PixelFormat pixelFormat, ImageFormat rawFormat)
+        {
+            var bitmap = new Bitmap(Helpers.GetTestBitmapPath(filename));
```
Updated

hughbe commented:
Thanks for reviewing. I know I probably bombarded you with PRs during your vacation, sorry!
I've updated PR feedback.

As per the disposal of test resources, you are right. It's not a massive problem as they will all be finalized, but I plan to submit a PR when the backlog is reduced fixing this. I've updated this PR accordingly, and will dispose in future PRs

stephentoub commented:
Overall looks good.  My only concern is the binary files.  @mellinoe, have we been checking in the binary files for testing?  For other projects, we've tried to put them in a separate test data repo to avoid polluting corefx with them.

hughbe commented:
We have binary files for IconTests already in the repo - they come directly from mono. As per test data repo, up to you

mellinoe commented:
The test files used here are so tiny that I'm okay with just including them in the repo. Some of the other test projects end up using hundreds of MB of test data, although I guess it's a bit of an arbitrary line to draw.

hughbe commented:
I've tried to fix the merge conflicts through the github UI to avoid you having to re-review :0

hughbe commented:
Test Windows x64 Debug Build

hughbe commented:
Rebased to fix merge conflicts!

stephentoub commented on src/System.Drawing.Common/tests/BitmapTests.cs:
```diff
@@ -0,0 +1,1674 @@
+            {
+                BitmapData data = bitmap.LockBits(new Rectangle(0, 0, 1, 1), ImageLockMode.ReadOnly, bitmap.PixelFormat);
+                data.Scan0 = IntPtr.Zero;
+
```
The serialization tests needs to be deleted.

hughbe replied:
Ah bummer, I didn't think to recompile the source before running the tests after I fixed the merge conflict
